Mathlib/Topology/Order/HullKernel.lean

```diff
@@ -0,0 +1,289 @@
+
+/- When `α` is complete, the relative basis for the Lower topology is also closed under arbitary
+unions.-/
+lemma sUnion_Ici_Compl_eq (S : Set α) : ⋃₀ { T ↓∩ (Ici a)ᶜ | a ∈ S } = T ↓∩ (Ici (sSup S))ᶜ := by
+  rw [Set.preimage_compl, ← sInter_Ici_eq, compl_sInter, sUnion_eq_compl_sInter_compl]
+  simp only [preimage_compl, sInter_image, mem_setOf_eq, iInter_exists, biInter_and',
+    iInter_iInter_eq_right, compl_compl, compl_iInter, sUnion_image, iUnion_exists, biUnion_and',
+    iUnion_iUnion_eq_right]
+
+/- When `α` is complete, a set is Lower topology relative-open if and only if it is of the form
+`T ↓∩ (Ici a)ᶜ` for some `a` in `α`.-/
+lemma isOpen_iff (S : Set T) : IsOpen S ↔ ∃ (a : α), S = T ↓∩ (Ici a)ᶜ := by
+  constructor
+  · intro h
+    let R := {a : α | T ↓∩ (Ici a)ᶜ ⊆ S}
+    use sSup R
+    rw [← sUnion_Ici_Compl_eq,
+      IsTopologicalBasis.open_eq_sUnion' (relativeLowerIsTopologicalBasis hT) h]
+    aesop
+  · intro h
+    cases' h with a ha
+    use (Ici a)ᶜ
+    constructor
+    · rw [isOpen_compl_iff]
+      exact isClosed_Ici
+    · rw [ha]
+
+/- When `α` is complete, a set is Lower topology relative-closed if and only if it is of the form
+`T ↓∩ (Ici a)` for some `a` in `α`.-/
+lemma isClosed_iff (S : Set T) : IsClosed S ↔ ∃ (a : α), S = T ↓∩ (Ici a) := by
+  rw [← isOpen_compl_iff, (isOpen_iff hT)]
+  constructor
+  · intros h
+    cases' h with a ha
+    use a
+    rw [preimage_compl, compl_inj_iff] at ha
+    exact ha
+  · intros h
+    cases' h with a ha
+    use a
+    rw [preimage_compl, compl_inj_iff]
+    exact ha
+
+/- The pair of maps `S → ⊓ S` (kernel) and `a → T ↓∩ (Ici a)` (hull) form an antitone Galois
+connection betwen the subsets of `T` and `α`. -/
+open OrderDual in
+theorem gc : GaloisConnection (α := Set T) (β := αᵒᵈ)
+    (fun S => toDual (sInf (S : (Set α)))) (fun a => T ↓∩ (Ici (ofDual a))) := fun S a => by
+  constructor
+  · intro h b hbS
+    rw [mem_preimage, mem_Ici]
+    rw [← ofDual_le_ofDual] at h
+    simp only [toDual_sInf, ofDual_sSup, le_sInf_iff, mem_preimage, ofDual_toDual, mem_image,
+      Subtype.exists, exists_and_right, exists_eq_right, forall_exists_index] at h
+    exact h _ (Subtype.coe_prop b) hbS
+  · intro h
+    simp only [toDual_sInf, sSup_le_iff, mem_preimage, mem_image, Subtype.exists, exists_and_right,
+      exists_eq_right, ← ofDual_le_ofDual, forall_exists_index, OrderDual.forall, ofDual_toDual]
+    exact fun b _ hbS => h hbS
+
+lemma gc_closureOperator_eq (S : Set T) : gc.closureOperator S = T ↓∩ (Ici (sInf S)) := by
+  simp only [toDual_sInf, GaloisConnection.closureOperator_apply, ofDual_sSup]
+  rw [← preimage_comp, ← OrderDual.toDual_symm_eq, Equiv.symm_comp_self, preimage_id_eq, id_eq]
+
+variable (T)
+
+/-- `T` is said to order generate `α` if, for every `a` in `α`, there exists a subset of `T` such
+that `a` is the inf of `S`. -/
+def OrderGenerate := ∀ (a : α), ∃ (S : Set T), a = sInf (S : Set α)
```

Comment:
```suggestion
def OrderGenerate := ∀ (a : α), ∃ S ⊆ T, a = sInf S
```
and everywhere else.

Reply:
`S ⊆ T` gives `S : Set α` rather than `S : Set T`.

Reply:
Yes, that's the point